File: src/py/Calendar.py
"""
Source Code   : Tugas2-1606918055.py
Author        : Wisnu Pramadhitya R
Created       : October 04 - Oktober 14, 2016
Description   : aplikasi Kalender sederhana yang mampu menampilkan kalender 
per bulan sesuai bulan dan tahun yang dimasukkan oleh pengguna.
"""
import sys

mdays = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
def leapyear(year):
    'Return True if leap year, False if not.'
    return year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)

def getWeekday(year):
    'Return calculation of total days from 1800-year\nday: int total days\nyear: int 1800-2099\n'
    day = 0
    for i in range(1800, year):
        if (leapyear(i)): day += 1 # +1 if the year is leap year
        day += 365
    return day

def getMonthday(year, month):
    'Return calculation of total days from 1 January -> month\nday: \
    int total days\nyear = int 1800-2099\nmonth: int 1-12'
    day = 0
    for i in range(1, month):
        if (i == 2 and leapyear(year)): day += 29 # +29 if the year is leap year and february
        else: day += mdays[i]
    return day

def calendar(year, month):
    'Return a simple Calendar as a text that already formatted' 
    nmonths = ('', 'Januari', 'Februari', 'Maret', 'April', 'Mei', 'Juni', \
            'Juli', 'Agustus', 'September', 'Oktober', 'November', 'Desember')

    if (leapyear(year)): mdays[2] = 29 
        # change the day to 29 on index mdays[2] means February if user input is leap year
    startday = (3 + getWeekday(year) + getMonthday(year, month)) % 7
        # calculate the initial day of a month
        # +3 is normalization from 1 january 1800, that was Wednesday (3)
    
    print('{} {}\n  M    S    S    R    K    J    S'.format(nmonths[month],year))
    for i in range(startday):
        print('    ', end=' ')
    for day in range(1, mdays[month]+1):
        print('{:3}'.format(day), end='  ')
        if ((day + startday) % 7 == 0 or day == mdays[month]): print()

def main():
    'main program'
    month, year = 0, 0
    while True:
        try:
            month = int(input('masukkan bulan 1-12 (-1 untuk keluar): '))
            if (month == -1): sys.exit()
            elif not (1 <= month <= 12): 
                print('Perhatian: input bulan 1-12, selain itu, error. Coba lagi!'); continue 
                # raise error
            year = int(input('masukkan tahun: '))
            if not (1800 <= year <= 2099): 
                print('Perhatian: input tahun 1800-2099, selain itu, error. Coba lagi!'); continue 
                # raise error
            
            calendar(year,month) # display calendar
        except (ValueError): print('ValueError: input harus berupa <int>, selain itu, error. Coba lagi!') 
        # raise error /ValueError/
    
if __name__ == "__main__":
    #main()
    calendar(2016,2)

File: src/java/Calendar.java
package ddp2.Calendar;

/** calendar 
 * show calendar by year and Month
 * print to display
*/
public class Calendar {
    // initiate
    private int month, year;
    private int[] mdays = {0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
    private String[] months = {" ", "January", "February", "March", "April", "May", "June",
            "July", "August", "September", "October", "November", "December"};

    public Calendar(int year) {
        this.year = year;
    }

    public Calendar(int year, int month) {
        this.month = month;
        this.year = year;
    }

    public void print(){
        // flag
        if (this.month == 0){
            // print all month in a year
            // calculate the day from 1800 to n year
            // + 3 for normalize the first day of January 1800
            int dayByYear = 3 + getDay();

            // change the February to 29 if detect the leap year
            if (leapYear()) this.mdays[2] = 29;

            // print calendar
            for (int i = 1; i <= 12; i++){
                // select month
                this.month = i;

                // calculate the first day
                int dayOne = (dayByYear + getMonthDay()) % 7;

                // print calendar format
                System.out.printf("%s %d\n", months[month], year);
                System.out.println(" S  M  T  W  T  F  S");

                // print the day
                for (int j = 0; j < dayOne; j++) {
                    System.out.print("   ");
                }
                for (int j = 1; j <= this.mdays[month]; j++) {
                    System.out.printf("%2d ", j);
                    // print as new week
                    if (((j + dayOne) % 7 == 0) || (j == this.mdays[month]))
                        System.out.println();
                }
                System.out.println();
            }
            // normalize
            this.month = 0;
        }

        else {
            // print by month and year
            // calculate the first day
            // + 3 for normalize the first day of January 1800
            int dayOne = (3 + getDay() + getMonthDay()) % 7;
            if (leapYear()) this.mdays[2] = 29;

            // print Calendar format
            System.out.format("%s %d\n", months[month], year);
            System.out.println(" S  M  T  W  T  F  S");

            // print the day
            for (int i = 0; i < dayOne; i++) {
                System.out.print("   ");
            }
            for (int i = 1; i <= this.mdays[month]; i++) {
                System.out.printf("%2d ", i);
                if (((i + dayOne) % 7 == 0) || (i == this.mdays[month])) System.out.println();
            }
        }
    }

    public boolean leapYear() {
        // return true if n year is a leapYear
        return (this.year % 4) == 0 && ((this.year % 100) != 0 || (this.year % 400) == 0);
    }

    private int getDay() {
        int day, i;

        day = 0;
        // calculate day from 1800 to n year
        for (i = 1800; i < this.year; i++) {
            if (leapYear()) day++;
            day += 365;
        }
        return day;
    }

    private int getMonthDay() {
        int day, i;

        day = 0;
        // calculate day from day 0 to the end of n month
        for (i = 1; i < this.month; i++) {
            if ((i == 2) && leapYear()) day++;
            else day += this.mdays[i];
        }
        return day;
    }
}